# 错误处理与数据返回



## 正常的路由处理

在`api/xxx.js`中进行路由的设置

- 进行请求参数的校验，分别进行`string`类型和`function`类型的校验
  - `string`类型：将错误存放在`this.errors.push(xxxxException)`
  - `function`类型：将错误存放在`this.errors.push(xxxxException)`
  - 检测完成后，检测`this.errors.length`，将错误进行`throw e`
- 进行业务逻辑的处理，如果处理成功，则直接`ctx.body=new SuccessException().getData()`
- 进行业务逻辑的处理，如果处理失败，则直接`ctx.body=new XXXXException().getData()`



## 全局的错误捕获处理

使用中间件放在最后一个位置，然后进行`try-catch`的错误捕获

- 如果拿到的是`HttpException`类型，则进行`ctx.body=new XXXXException().getData()`返回
- 如果拿到不是`HttpException`类型，则进行`ctx.body={message: "未知错误"}`返回


## throw错误的场景

- 进行请求参数的校验的方法，也就是`src/lib/common-validator.js`的`validate()`进行统一的`throw e`
- 业务逻辑中，比如校验token中发现token已经过期，直接进行`throw e`，中断程序的继续执行


